packages/jest-haste-map/src/index.js

```
@@ -0,0 +1,370 @@
+    const mocksPattern = this._options.mocksPattern;
+    const promises = [];
+    const setModule = (id, module) => {
+      if (!map[id]) {
+        map[id] = Object.create(null);
+      }
+      const moduleMap = map[id];
+      const platform =
+        getPlatformExtension(module[H.PATH]) || H.GENERIC_PLATFORM;
+      const existingModule = moduleMap[platform];
+      if (existingModule && existingModule[H.PATH] !== module[H.PATH]) {
+        console.warn(
+          `@providesModule naming collision:\n` +
+          `  Duplicate module name: ${id}\n` +
+          `  Paths: ${module[H.PATH]} collides with ` +
+          `${existingModule[H.PATH]}\n\n` +
+          `This warning is caused by a @providesModule declaration ` +
+          `with the same name accross two different files.`
+        );
+        return;
+      }
+
+      moduleMap[platform] = module;
+    };
+
+    for (const filePath in hasteMap.files) {
+      if (mocksPattern && mocksPattern.test(filePath)) {
+        mocks[path.basename(filePath, path.extname(filePath))] = filePath;
+      }
+
+      const fileMetadata = hasteMap.files[filePath];
+      const moduleMetadata = hasteMap.map[fileMetadata[H.ID]];
+      if (fileMetadata[H.VISITED]) {
+        if (!fileMetadata[H.ID]) {
+          continue;
+        } else if (fileMetadata[H.ID] && moduleMetadata) {
+          map[fileMetadata[H.ID]] = moduleMetadata;
+          continue;
+        }
+      }
+
+      promises.push(
+        this._getWorker()({filePath}).then(
+          metadata => {
+            fileMetadata[H.VISITED] = 1;
+            if (metadata.id) {
+              fileMetadata[H.ID] = metadata.id;
+              setModule(metadata.id, metadata.module);
+            }
+            fileMetadata[H.DEPENDENCIES] = metadata.dependencies || [];
+          },
+          error => {
+            // If a file cannot be read we remove it from the file list and
+            // ignore the failure silently.
```

I believe watchman's `exists` flag on files might not always be entirely correct. I chose to not trust watchman here and just ignore errors silently. We might want to log or throw at some point but it should be good enough for now.